### PR TITLE
feature/#106-Support-Sqlalchemy-1.4-part-1

### LINKED ISSUE #106: Support Sqlalchemy 1.4
sqlalchemy_exasol python package is not compatible with new version of sqlalchemy 1.4+ when used with turbodbc dialect. Below is the code to reproduce this issue.

```python

from sqlalchemy import create_engine
 
exa_engine = create_engine("exa+turbodbc://sys:exasol@exasoldb:8888")
exa_engine.connect()
 
 
################################### ERROR #########################################
 
 
    exec(code, run_globals)
  File "/app/test.py", line 6, in <module>
    exa_engine.connect()
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 3166, in connect
    return self._connection_cls(self, close_with_result=close_with_result)
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 96, in __init__
    else engine.raw_connection()
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 3245, in raw_connection
    return self._wrap_pool_connect(self.pool.connect, _connection)
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 3212, in _wrap_pool_connect
    return fn()
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 301, in connect
    return _ConnectionFairy._checkout(self)
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 761, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 419, in checkout
    rec = pool._do_get()
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/pool/impl.py", line 145, in _do_get
    self._dec_overflow()
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/pool/impl.py", line 142, in _do_get
    return self._create_connection()
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 247, in _create_connection
    return _ConnectionRecord(self)
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 362, in __init__
    self.__connect()
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 616, in __connect
    pool.dispatch.connect.for_modify(
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/event/attr.py", line 329, in _exec_w_sync_on_first_run
    self(*args, **kw)
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/event/attr.py", line 343, in __call__
    fn(*args, **kw)
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py", line 1676, in go
    return once_fn(*arg, **kw)
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/engine/create.py", line 675, in first_connect
    dialect.initialize(c)
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 394, in initialize
    self.server_version_info = self._get_server_version_info(
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy_exasol/turbodbc.py", line 82, in _get_server_version_info
    result = connection.execute(query).fetchone()[0].split('.')
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1247, in execute
    return self._exec_driver_sql(
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1546, in _exec_driver_sql
    ret = self._execute_context(
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1696, in _execute_context
    context.pre_exec()
  File "/root/venv/lib/python3.8/site-packages/sqlalchemy_exasol/base.py", line 336, in pre_exec
    if self.isdelete or self.root_connection.dialect.server_version_info < (4, 1, 8):
TypeError: '<' not supported between instances of 'NoneType' and 'tuple'

```

### PR DESCRIPTION
**Pull request by Nicoretti**
- Update sqlalchemy dependency to base version to 1.4.0
- Update ssl/tls test to comply with new sqla 1.4 api
- Remove backwards compatibility with SQLA < 1.4 for ssl/tls test
- Add changelog entry
- Disable statement cache for pyodbc and turbodbc dialects
- Upgrade tests to use new introspection api
- Update URL creation to new api
- Fix pre_exec edge case
- Restructure test setup & execution

### CODE REVIEW

File: sqlalchemy_exasol/base.py [hunk 1]
```diff
@@ -762,48 +762,50 @@ def pre_exec(self):
...
-                        )
-                    elif isinstance(value, str):
-                        db_query = db_query.replace(ident, "'%s'" % value, 1)
-                    else:
-                        raise TypeError("Data type not supported: %s" % type(value))
-            self.statement = db_query
-            self.parameters = [[]]
+        server_version = self.root_connection.dialect.server_version_info
+        skip_pre_exec = (
+            not self.isdelete and server_version is None or server_version >= (4, 1, 8)
```
Comment (tkilias) [inline sqlalchemy_exasol/base.py:769]: I think, it is save to drop support for Exasol version 4

File: changelog.d/20220810_120055_nicola.coretti__106_Support_Sqlalchemy_1_4.rst [hunk 2]
```diff
@@ -0,0 +1,10 @@
+🔧 Changed
+----------
+- Update SQLAlchemy dependency to >= 1.4
+
+    .. warning::
+
+        This may impact the performance (see also `SQLAlchemy docs <https://docs.sqlalchemy.org/en/14/faq/performance.html#why-is-my-application-slow-after-upgrading-to-1-4-and-or-2-x>`_).
```
Comment (tkilias) [inline changelog.d/20220810_120055_nicola.coretti__106_Support_Sqlalchemy_1_4.rst:7]: If, that means probably quite a lot work

File: pyproject.toml [hunk 3]
```diff
@@ -54,13 +54,14 @@ exclude = []
...
 python = ">=3.8,<4.0"
-SQLAlchemy = ">=1.3.24,<1.4"
+SQLAlchemy = ">=1.4"
 pyodbc = ">=4.0.34"
 packaging = "^21.3"
 # TODO: Reenable once the causing issues got fixed:
 #  - https://github.com/blue-yonder/turbodbc/issues/358
 #  - https://github.com/exasol/sqlalchemy-exasol/issues/146
 # turbodbc = { version = ">4.5.3", optional = true }
+turbodbc = {path = "turbodbc-4.5.3.tar.gz"}
```
Comment (tkilias) [inline pyproject.toml]: This needs to go later

File: changelog.d/20220810_120055_nicola.coretti__106_Support_Sqlalchemy_1_4.rst (hunk 2 above)
Comment (tkilias) [reply changelog.d/20220810_120055_nicola.coretti__106_Support_Sqlalchemy_1_4.rst:7]: Do you know, if it at least caches the metadata queries?

File: sqlalchemy_exasol/base.py [hunk 4]
```diff
@@ -762,48 +762,50 @@ def pre_exec(self):
...
-                        db_query = db_query.replace(
-                            ident, "'%s'" % value.decode("UTF-8"), 1
-                        )
-                    elif isinstance(value, str):
-                        db_query = db_query.replace(ident, "'%s'" % value, 1)
-                    else:
-                        raise TypeError("Data type not supported: %s" % type(value))
-            self.statement = db_query
-            self.parameters = [[]]
+        server_version = self.root_connection.dialect.server_version_info
```
Comment (tkilias) [inline sqlalchemy_exasol/base.py:765]: Did you change anything here? Or, is this reformating? Because the diff is unfortunately not really helpful

File: sqlalchemy_exasol/base.py (hunk 1 above)
Comment (Nicoretti) [reply sqlalchemy_exasol/base.py:769]: nice, still we probably still need it for the delete case, at least that needs to be clarified.

File: changelog.d/20220810_120055_nicola.coretti__106_Support_Sqlalchemy_1_4.rst (hunk 2 above)
Comment (Nicoretti) [reply changelog.d/20220810_120055_nicola.coretti__106_Support_Sqlalchemy_1_4.rst:7]: I don't no yet. Because the caching topic seems to be quite complex I moved it into an follow up issue, so we can focus on the migration to 1.4 first.

File: pyproject.toml (hunk 3 above)
Comment (Nicoretti) [reply pyproject.toml]: Yes, good reminder. But need it for testing right now (at least until https://github.com/blue-yonder/turbodbc/pull/359 gets merged and released).

File: sqlalchemy_exasol/base.py (hunk 4 above)
Comment (Nicoretti) [reply sqlalchemy_exasol/base.py:765]: no I just added a extra variable to store the `server_version` so the code using it will be more concise.

**Review by umitbuyuksahin — DISMISSED** [review]

File: sqlalchemy_exasol/merge.py [hunk 5]
```diff
@@ -11,115 +11,136 @@
...
 
-    def _get_source_cols(self):
-        source_cols = {}
-        elements_to_check = list(self._source_expr.get_children())
-        for e in elements_to_check:
-            if isinstance(e, ColumnClause):
-                source_cols[e.name] = e
-        return source_cols
+    def _source_columns(self):
+        elements = list(
```
Comment (tkilias) [inline sqlalchemy_exasol/merge.py:43]: why do you use list + iter comprehension instead by a list comprehension

File: sqlalchemy_exasol/merge.py [hunk 6]
```diff
@@ -11,115 +11,136 @@
...
-            for c in self._target_table.c:
-                if c not in self._on_columns and c.name in source_cols:
-                    values[c] = source_cols[c.name]
+            src_columns = self._source_columns()
+            columns = (
+                column
+                for column in self._target_table.c
+                if column not in self._on_columns and column.name in src_columns
+            )
+            [values.update({column: src_columns[column.name]}) for column in columns]
```
Comment (tkilias) [inline sqlalchemy_exasol/merge.py]: you probably could create values in on dict comprehension with two for's, but this is not always better to read, but you can likely use a dict comprehension instead of values update.

File: sqlalchemy_exasol/merge.py [hunk 7]
```diff
@@ -11,115 +11,136 @@
...
         self._on = on
-        self._on_columns = []
-        elements_to_check = list(on.get_children())
-        for e in elements_to_check:
-            if isinstance(e, Column):
-                if e.table == self._target_table:
-                    self._on_columns.append(e)
-            else:
-                elements_to_check.extend(e.get_children())
+        self._on_columns = list(
```
Comment (tkilias) [inline sqlalchemy_exasol/merge.py:29]: I think, the logic doesn't equate the previous version, becaue the previous version looks like a serialized recursive descent into a tree

File: sqlalchemy_exasol/merge.py [hunk 8]
```diff
@@ -11,115 +11,136 @@
...
-            source_cols = self._get_source_cols()
             values = {}
-            for c in self._target_table.c:
-                if c.name in source_cols:
-                    values[c] = source_cols[c.name]
+            src_columns = self._source_columns()
+            columns = (
+                column for column in self._target_table.c if column.name in src_columns
+            )
+            [values.update({column: src_columns[column.name]}) for column in columns]
```
Comment (tkilias) [inline sqlalchemy_exasol/merge.py]: you probably could create values in on dict comprehension with two for's, but this is not always better to read, but you can likely use a dict comprehension instead of values update.

File: sqlalchemy_exasol/merge.py [hunk 9]
```diff
@@ -11,115 +11,136 @@
...
+        sql = "MERGE INTO %s " % compiler.process(self._target_table, asfrom=True)
+        sql += "USING %s " % compiler.process(self._source_expr, asfrom=True)
+        sql += "ON ( %s ) " % compiler.process(self._on)
+        if self._is_merge_update():
+            sql += self._compile_update(compiler)
+        if self._is_merge_delete():
+            sql += self._compile_delete(compiler)
+        if self._is_merge_insert():
+            sql += self._compile_insert(compiler)
+        return sql
```
Comment (tkilias) [inline sqlalchemy_exasol/merge.py]: I guess, you simply extracted the parts into methods, this way better understandable

**Review by tkilias — CHANGES_REQUESTED** [review]

File: sqlalchemy_exasol/base.py (hunk 1 above)
Comment (tkilias) [reply sqlalchemy_exasol/base.py:769]: yep, think the same

File: .github/workflows/ci-cd.yml [hunk 10]
```diff
@@ -49,79 +82,74 @@ jobs:
...
+          python-version: ${{ matrix.python }}
 
-    - name: Install poetry
-      uses: abatilo/actions-poetry@v2.1.5
-      with:
-        poetry-version: 1.1.13
+      - name: Install poetry
+        uses: abatilo/actions-poetry@v2.1.5
+        with:
+          poetry-version: 1.1.13
```
Comment (tkilias) [inline .github/workflows/ci-cd.yml]: Poetry 1.2.0

File: .github/workflows/ci-cd.yml [hunk 11]
```diff
@@ -49,79 +82,74 @@ jobs:
...
```
Comment (tkilias) [inline .github/workflows/ci-cd.yml]: Poetry 1.2.0

File: .github/workflows/ci.yaml [hunk 12]
```diff
@@ -36,9 +36,42 @@ jobs:
...
+
+      - name: Setup Python ${{ matrix.python }}
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python }}
+
+      - name: Install poetry
+        uses: abatilo/actions-poetry@v2.1.5
+        with:
+          poetry-version: 1.1.13
```
Comment (tkilias) [inline .github/workflows/ci.yaml]: Poetry 1.2.0

File: .github/workflows/ci.yaml [hunk 13]
```diff
@@ -53,72 +86,71 @@ jobs:
...
+          python-version: ${{ matrix.python }}
 
-    - name: Install poetry
-      uses: abatilo/actions-poetry@v2.1.5
-      with:
-        poetry-version: 1.1.13
+      - name: Install poetry
+        uses: abatilo/actions-poetry@v2.1.5
+        with:
+          poetry-version: 1.1.13
```
Comment (tkilias) [inline .github/workflows/ci.yaml]: Poetry 1.2.0

File: .github/workflows/ci.yaml [hunk 14]
```diff
@@ -53,72 +86,71 @@ jobs:
...
-      working-directory: ..
-      env:
-        ITDE_URL: "https://github.com/exasol/integration-test-docker-environment.git"
-        ITDE_TAG: "0.11.0"
+      - name: Checkout test environment
+        run: git clone --depth 1 --branch ${ITDE_TAG} ${ITDE_URL}
+        working-directory: ..
+        env:
+          ITDE_URL: "https://github.com/exasol/integration-test-docker-environment.git"
+          ITDE_TAG: "0.11.0"
```
Comment (tkilias) [inline .github/workflows/ci.yaml:121]: Use ITDE 1.1.0

File: .github/workflows/ci-cd.yml [hunk 15]
```diff
@@ -49,79 +82,74 @@ jobs:
...
-      working-directory: ..
-      env:
-        ITDE_URL: "https://github.com/exasol/integration-test-docker-environment.git"
-        ITDE_TAG: "0.11.0"
+      - name: Checkout test environment
+        run: git clone --depth 1 --branch ${ITDE_TAG} ${ITDE_URL}
+        working-directory: ..
+        env:
+          ITDE_URL: "https://github.com/exasol/integration-test-docker-environment.git"
+          ITDE_TAG: "0.11.0"
```
Comment (tkilias) [inline .github/workflows/ci-cd.yml:117]: Use 1.1.0

File: doc/developer_guide/integration_tests.rst [hunk 16]
```diff
@@ -1,5 +1,41 @@
-Integration Test Setup
-======================
+Integration Test
+================
+
+The integration tests are located within `test/integration`. They are split in
+two groups.
+
+#. The sqlachmey conformance test suite
```
Comment (tkilias) [inline doc/developer_guide/integration_tests.rst]: SQLAlchemy is a named entity and probably isn't written lower case

File: doc/developer_guide/integration_tests.rst [hunk 17]
```diff
@@ -1,5 +1,41 @@
...
+
+    The sqlalchemy conformance test suite is provided and maintained by the sqlalchemy project and intended to support third party dialect developers.
+    For further details see also `README.dialects.rst <https://docs.exasol.com/db/latest/sql_reference.htm>`_.
+
+#. Our custom exasol test suite
```
Comment (tkilias) [inline doc/developer_guide/integration_tests.rst]: Exasol is a named entity and probably isn't written lower case

File: doc/developer_guide/integration_tests.rst [hunk 18]
```diff
@@ -1,5 +1,41 @@
...
+
+.. note::
+
+    In order to reduce the likelihood of test side effects, the `sqlalchemy` and the `exasol` test suites
+    are executed in separate pytest test runs.
+
+.. attention::
+
+    The exasol database does do implicit schema/context changes (open & close schema)
+    in certain scenarios. Keep this in mind when setting up (writing) tests.
```
Comment (tkilias) [inline doc/developer_guide/integration_tests.rst:24]: Using a new connection can avoid aide effects.

File: doc/developer_guide/integration_tests.rst [hunk 19]
```diff
@@ -1,5 +1,41 @@
...
+        When you create a new schema, you implicitly open this new schema.
+
+    #. DROP SCHEMA sometimes switches the context to <null>
+
+        If the CURRENT_SCHEMA is dropped an implicit context switch to <null> is done
+
+    For further details have a look at the `Exasol-Documentation <https://docs.exasol.com/db/latest/sql_reference.htm>`_.
+
+
+Setup
```
Comment (tkilias) [inline doc/developer_guide/integration_tests.rst:42]: With setup, I had thought you mean the local test setup and not only the GitHub actions

File: doc/developer_guide/integration_tests.rst [hunk 20]
```diff
@@ -1,5 +1,41 @@
...
+=====
 
 Integration testing is done by GitHub Actions contained in this repository, which provide a CI/CD pipeline to test, build, and deploy sqlalchemy_exasol to Pypi.
```
Comment (tkilias) [inline doc/developer_guide/integration_tests.rst]: PyPI is also a named entity and as this it should written as it is on their website.

File: noxfile.py [hunk 21]
```diff
@@ -230,20 +115,26 @@ def is_version_in_sync() -> bool:
     session.notify("code-format")
     session.notify("type-check")
     session.notify("lint")
-    session.notify("type-check")
```
Comment (tkilias) [inline noxfile.py:233]: Did you remove mypy. Is there a specific reason?

File: noxfile.py [hunk 22]
```diff
@@ -287,12 +179,13 @@ def populate() -> None:
                 connection,
                 (
                     "CREATE SCHEMA TEST_SCHEMA;",
-                    "CREATE SCHEMA TEST_SCHEMA_2;",
+                    # "CREATE SCHEMA TEST_SCHEMA_2;",
```
Comment (tkilias) [inline noxfile.py:182]: If we don't need this, then remove it.

File: noxfile.py [hunk 23]
```diff
@@ -303,89 +196,148 @@ def stop_db(session: Session) -> None:
...
-    with odbcconfig() as (config, env):
-        uri = "".join(
-            [
-                "exa+{connector}:",
-                "//sys:exasol@localhost:{db_port}",
-                "/TEST?CONNECTIONLCALL=en_US.UTF-8&DRIVER=EXAODBC&SSLCertificate=SSL_VERIFY_NONE",
-            ]
-        ).format(connector=connector, db_port=Settings.DB_PORT)
-        session.run("pytest", "--dropfirst", "--dburi", uri, external=True, env=env)
+    Attention:
```
Comment (tkilias) [inline noxfile.py:204]: We should later maybe codify this condition

File: .github/workflows/ci.yaml [hunk 24]
```diff
@@ -53,72 +86,71 @@ jobs:
...
+          python-version: ${{ matrix.python }}
 
-    - name: Install poetry
-      uses: abatilo/actions-poetry@v2.1.5
-      with:
-        poetry-version: 1.1.13
+      - name: Install poetry
+        uses: abatilo/actions-poetry@v2.1.5
+        with:
+          poetry-version: 1.1.13
```
Comment (tkilias) [inline .github/workflows/ci.yaml:150]: Poetry 1.2.0

File: noxfile.py [hunk 25]
```diff
@@ -303,89 +196,148 @@ def stop_db(session: Session) -> None:
...
+@nox.session(python=False, name="clean-docs")
+def clean(session: Session) -> None:
+    """Remove all documentation artifacts"""
+    if DOC_BUILD.exists():
+        rmtree(DOC_BUILD.resolve())
+        session.log(f"Removed {DOC_BUILD}")
+
+
+@nox.session(python=False, name="build-docs")
+def build(session: Session) -> None:
```
Comment (tkilias) [inline noxfile.py:318]: build-docs, build ja not specific enough and could be confused with building then project

File: noxfile.py [hunk 26]
```diff
@@ -303,89 +196,148 @@ def stop_db(session: Session) -> None:
...
+    )
+    session.notify(
+        find_session_runner(session, f"exasol-tests"),
+        posargs=["--connector", f"{args.connector}"],
+    )
+    session.notify(find_session_runner(session, "db-stop"))
+
+
+@nox.session(python=False, name="clean-docs")
+def clean(session: Session) -> None:
```
Comment (tkilias) [inline noxfile.py:310]: clean_docs because clean could be confused

File: noxfile.py [hunk 27]
```diff
@@ -423,29 +375,119 @@ def create_parser() -> ArgumentParser:
...
+                    "--output",
+                    f"skipped-tests-{connector}.csv",
+                    f"{connector}",
+                    f"{report}",
+                )
+
+
+@nox.session(name="check-links", python=False)
+def check_links(session: Session) -> None:
+    """Checks weather or not all links in the documentation can be accessed"""
```
Comment (tkilias) [inline noxfile.py:475]: I thought sphinx takes care of this

File: pyproject.toml [hunk 28]
```diff
@@ -87,8 +87,7 @@ turbodbc = ["turbodbc"]
 "exa.turbodbc" = "sqlalchemy_exasol.turbodbc:EXADialect_turbodbc"
 
 [tool.pytest.ini_options]
-addopts= "--tb native -v -r fxX"
-python_files= "test/*test_*.py"
+addopts= "--tb native -v -r fxX --log-debug=sqlalchemy.engine --log-debug=sqlalchemy.pool"
```
Comment (tkilias) [inline pyproject.toml:90]: Do we have a description what these parameters mean?

File: setup.cfg [hunk 29]
```diff
@@ -1,3 +1,7 @@
+[db]
+exasol-pyodbc=exa+pyodbc://sys:exasol@localhost:8888/TEST?CONNECTIONCALL=en_US.UTF-8&DRIVER=EXAODBC&SSLCertificate=SSL_VERIFY_NONE
```
Comment (tkilias) [inline setup.cfg:2]: A very good, you are setting the locale to a defined value

File: sqlalchemy_exasol/merge.py [hunk 30]
```diff
@@ -1,3 +1,8 @@
+# Refactoring notes
```
Comment (tkilias) [inline sqlalchemy_exasol/merge.py:1]: Are these still needed?

File: sqlalchemy_exasol/merge.py [hunk 31]
```diff
@@ -11,115 +16,140 @@
...
-                if c not in self._on_columns and c.name in source_cols:
-                    values[c] = source_cols[c.name]
-        self._merge_update_values = ValuesBase(self._target_table, values, [])
+            src_columns = self._source_columns()
+            columns = (
+                column
+                for column in self.table.c
+                if column not in self._on_columns and column.name in src_columns
+            )
+            [values.update({column: src_columns[column.name]}) for column in columns]
```
Comment (tkilias) [inline sqlalchemy_exasol/merge.py:63]: List comprehension should not be used , when the list is not assigned. In that case a normal for loop is usually recommended. An alternative in this case might be a dict or set merge in case values is a dict or set.

File: sqlalchemy_exasol/merge.py [hunk 32]
```diff
@@ -11,115 +16,140 @@
...
         if values is None:
-            source_cols = self._get_source_cols()
             values = {}
-            for c in self._target_table.c:
-                if c.name in source_cols:
-                    values[c] = source_cols[c.name]
-        self._merge_insert_values = ValuesBase(self._target_table, values, [])
+            src_columns = self._source_columns()
+            columns = (column for column in self.table.c if column.name in src_columns)
+            [values.update({column: src_columns[column.name]}) for column in columns]
```
Comment (tkilias) [inline sqlalchemy_exasol/merge.py:80]: Same as above

File: sqlalchemy_exasol/requirements.py [hunk 33]
```diff
@@ -198,3 +198,33 @@ def standalone_null_binds_whereclause(self):
         WHERE clause, in situations where it has to be typed.
         """
         return exclusions.closed()
+
```
Comment (tkilias) [inline sqlalchemy_exasol/requirements.py:201]: I guess this file, was already at this path. However, it actually has to do with tests.

**Review by tkilias — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml (hunk 14 above)
Comment (Nicoretti) [reply .github/workflows/ci.yaml:121]: I rather would switch versions of ITDE at the end (once all tests are green) to reduce the amount of moving parts regarding testing.

File: .github/workflows/ci-cd.yml (hunk 15 above)
Comment (Nicoretti) [reply .github/workflows/ci-cd.yml:117]: Due to the fact that this potentially impacts testing itself I rather would pros pone this until all test are green again.

File: sqlalchemy_exasol/requirements.py (hunk 33 above)
Comment (Nicoretti) [reply sqlalchemy_exasol/requirements.py:201]: This must be part of the dialect itself (at least according to the [documentation](https://github.com/sqlalchemy/sqlalchemy/blob/main/README.dialects.rst#dialect-layout))

File: pyproject.toml (hunk 28 above)
Comment (Nicoretti) [reply pyproject.toml:90]: This are options documented and provided by the sqla pytest extension. Would you be in favor of removing them from the general settings?

File: .github/workflows/ci.yaml (hunk 14 above)
Comment (tkilias) [reply .github/workflows/ci.yaml:121]: ok, fine by me, maybe create a ticket as reminder

File: sqlalchemy_exasol/merge.py (hunk 30 above)
Comment (Nicoretti) [reply sqlalchemy_exasol/merge.py:1]: I would mostly ignore this file for now, with the changed version it needs a complete rework. I can roll it back to its original state though until then. What do you think?

File: pyproject.toml (hunk 28 above)
Comment (tkilias) [reply pyproject.toml:90]: I am find with them for this project, I only would add a comment where they are from

File: sqlalchemy_exasol/requirements.py (hunk 33 above)
Comment (tkilias) [reply sqlalchemy_exasol/requirements.py:201]: ok, I thought so also first, because it is somewhat a description of capabilities, was only surprised by the testing imports

File: sqlalchemy_exasol/merge.py (hunk 30 above)
Comment (tkilias) [reply sqlalchemy_exasol/merge.py:1]: ah ok, in that case, I think you can keep the refactoring, because it is way better to understand than the original and might make it easier to rewrite it.

File: noxfile.py (hunk 23 above)
Comment (tkilias) [reply noxfile.py:204]: ah, I think this one is outdated, I saw you call them seperately

File: pyproject.toml (hunk 28 above)
Comment (Nicoretti) [reply pyproject.toml:90]: Created a ticket https://github.com/exasol/sqlalchemy-exasol/issues/219
